apps/labs/posts/unlocking-c-level-performance-in-df-apply.md

```diff
@@ -0,0 +1,466 @@
+
+First, let's take a look at the speed of our apply function when raw is False, to provide a baseline.
+
+```python
+# Now on the raw data (ndarrays instead of Series)
+raw_f = lambda x: (x-x.mean()) / x.std()
+%timeit df.apply(f, engine="numba", axis=1, raw=False)
+```
+
+    326 ms ± 38.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
+
+And now, the raw performance.
+
+Notice as with before how operating on raw values is much faster than operating on the Series itself.
+
+```python
+%timeit df.apply(raw_f, engine="numba", axis=1, raw=True, engine_kwargs={"parallel": False})
+```
+
+    1.95 ms ± 115 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
+
+Now, let's trying operating in parallel.
+
+Here, we can see roughly a 5x speedup over just using raw.
+(Again, this was performed on my 2019 Intel MacBook Pro with 6 cores.)
+
+```python
+%timeit df.apply(raw_f, engine="numba", axis=1, raw=True, engine_kwargs={"parallel": True})
+```
+
+    268 µs ± 104 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
+
+Finally, for comparison's sake, let's look at the Python performance with raw=True, and also the
+time it takes for the vectorized equivalents.
+
+```python
+%timeit df.sub(df.mean(axis=1), axis=0).div(df.std(ddof=0, axis=1), axis=0)
+```
+
+    2.51 ms ± 30.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
+
+In this case, parallel apply with raw=True provided a good speedup, however,
+this is not guaranteed to happen in all cases.
+
+A good rule of thumb to follow is that if there already exists a numpy/pandas function
+that does what you want to do already, you should use that, as it is hard for numba
+to beat the optimized low-level routines in both of those libraries.
+
+However, if your operation chains together a lot of these operations, numba may provide a pretty
+good speedup. For example, in our normalization example above, we took the mean, did a subtraction,
+took the standard deviation, and then did a division. Because numpy evaluates each one of these operations
+eagerly, it may miss out on performance optimizations that the numba compiler is able to use, explaining the
+performance improvement with numba above.
+
+```python
+# Let's take a look at the performance of just a mean operation with numba
+# (both raw mode and parallel on)
+# and just with numpy/pandas
+
+f = lambda x: x.mean()
+%timeit df.apply(f, engine='numba', raw=True, engine_kwargs={'parallel': True})
+%timeit df.mean(axis=1)
+%timeit df.values.mean(axis=1)
+```
+
+    178 µs ± 60.9 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
+    1.1 ms ± 99.8 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
+    32.9 µs ± 2.74 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
+
+Here, we can see that although numba beats the pandas mean function (it's a bit slower than the numpy version because it does some extra work),
+it loses to the numpy mean function, illustrating the point above.
+
```

Comment:
You could consider if it's worth adding a graph, a la https://github.com/rgommers/explore-array-computing/blob/master/explore_xnd.ipynb? I quite like such log-log plots, they're visually appealing and it makes the scaling with data size (here cols and/or rows) easy to understand.

Reply:
Either way, one visual would be useful to have. In the presentation you showed me for the Pandas developer meeting you had some Snakeviz profiles, did you consider including one of those?

Reply:
The log-log plot makes sense to me. That's probably the most relevant for end users.

I'll probably not include a snakeviz profile, since even though they're helpful to look at, they don't have much meaning for non pandas devs.